### PR TITLE
Add 3D Touch to Care Card and Connect

### PR DESCRIPTION
**Pull request by tktsubota**
This adds peek and pop functionality to `OCKCareCardViewController` and `OCKConnectViewController`. When applying pressure to the table view cells, the peek view will show the respective detail view controller.

`OCKConnectDetailViewController` now leverages preview action items. When peeking into an `OCKConnectDetailViewController` from an `OCKConnectViewController`, swiping up will provide options to call, text, email, or send reports to the contact. Only the available options will be shown.

In order to do this, I added `UIViewControllerPreviewingDelegate` to `OCKCareCardViewController` and `OCKConnectViewController`. I implemented the `previewActionItems` method on `OCKConnectDetailViewController` to show the peek options.

To prevent a weird animation from occurring when applying pressure to the header in the Care Card, I changed `OCKCareCardTableViewHeader`'s superclass to be `UITableViewHeaderFooterView` and made slight modifications to preserve the blurring effect.

I also changed some helper methods to allow presentation of message and mail view controllers from both `OCKConnectViewController` and `OCKConnectDetailViewController`.

### CODE REVIEW

**Comment by umerkhan-apple** [comment]
@tktsubota It was a pleasure meeting you at WWDC. Thank you for this PR! Looking forward to testing out the 3D Touch additions. Could you resolve the merge conflicts on the branch?

**Comment by tktsubota** [comment]
@umerkhan-apple Yes, I'll get right to it.

**Comment by tktsubota** [comment]
@umerkhan-apple Conflict resolved now.

**Comment by tktsubota** [comment]
@umerkhan-apple Any updates on this?

**Comment by YuanZhu-apple** [comment]
@umerkhan-apple Can you review this?

File: CareKit/Connect/OCKConnectViewController.h [hunk 1]
```diff
@@ -50,7 +50,7 @@ NS_ASSUME_NONNULL_BEGIN
  @param contact                     The contact that is currently displayed.
  @param sourceView                  Source view can be used to present a popover on iPad.
  */
-- (void)connectViewController:(OCKConnectViewController *)connectViewController didSelectShareButtonForContact:(OCKContact *)contact presentationSourceView:(UIView *)sourceView;
+- (void)connectViewController:(OCKConnectViewController *)connectViewController didSelectShareButtonForContact:(OCKContact *)contact presentationSourceView:(nullable UIView *)sourceView;
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectViewController.h]: Update the sample app to reflect the change for nullability.

If you build the sample app, Xcode will automatically recommend the appropriate change.

File: CareKit/CareCard/OCKCareCardViewController.m [hunk 2]
```diff
@@ -497,4 +512,26 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
     return cell;
 }
```
Comment (umerkhan-apple) [inline CareKit/CareCard/OCKCareCardViewController.m]: Add one more new line.

File: CareKit/CareCard/OCKCareCardViewController.m [hunk 3]
```diff
@@ -497,4 +512,26 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
 
+#pragma mark - UIViewControllerPreviewingDelegate
+
+- (nullable UIViewController *)previewingContext:(id <UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location {
```
Comment (umerkhan-apple) [inline CareKit/CareCard/OCKCareCardViewController.m]: Remove nullable

File: CareKit/Connect/OCKConnectDetailViewController.m [hunk 4]
```diff
@@ -181,11 +182,11 @@ - (void)contactInfoTableViewCellDidSelectConnection:(OCKContactInfoTableViewCell
             break;
             
         case OCKConnectTypeMessage:
-            [self sendMessageToNumber:cell.contact.messageNumber.stringValue];
+            [self sendMessageToNumber:cell.contact.messageNumber.stringValue withDelegate:self usingPresentingViewController:self];
             break;
             
         case OCKConnectTypeEmail:
-            [self sendEmailToAddress:cell.contact.emailAddress];
+            [self sendEmailToAddress:cell.contact.emailAddress withDelegate:self usingPresentingViewController:self];
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectDetailViewController.m]: Is there a reason for making the presenting view controller a parameter?
Could you instead just use self inside the method?

File: CareKit/Connect/OCKConnectDetailViewController.m [hunk 5]
```diff
@@ -153,21 +154,21 @@ - (void)makeCallToNumber:(NSString *)number {
...
 
-- (void)sendEmailToAddress:(NSString *)address {
+- (void)sendEmailToAddress:(NSString *)address withDelegate:(id<MFMailComposeViewControllerDelegate>)emailDelegate usingPresentingViewController:(UIViewController *)presentingViewController {
     MFMailComposeViewController *emailViewController = [MFMailComposeViewController new];
     if ([MFMailComposeViewController canSendMail]) {
-        emailViewController.mailComposeDelegate = self;
+        emailViewController.mailComposeDelegate = emailDelegate;
         [emailViewController setToRecipients:@[address]];
-        [self presentViewController:emailViewController animated:YES completion:nil];
+        [presentingViewController presentViewController:emailViewController animated:YES completion:nil];
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectDetailViewController.m]: Can you use self here?

File: CareKit/Connect/OCKConnectDetailViewController.m [hunk 6]
```diff
@@ -153,21 +154,21 @@ - (void)makeCallToNumber:(NSString *)number {
...
 
-- (void)sendMessageToNumber:(NSString *)number {
+- (void)sendMessageToNumber:(NSString *)number withDelegate:(id<MFMessageComposeViewControllerDelegate>)messageDelegate usingPresentingViewController:(UIViewController *)presentingViewController {
     MFMessageComposeViewController *messageViewController = [MFMessageComposeViewController new];
     if ([MFMessageComposeViewController canSendText]) {
-        messageViewController.messageComposeDelegate = self;
+        messageViewController.messageComposeDelegate = messageDelegate;
         messageViewController.recipients = @[number];
-        [self presentViewController:messageViewController animated:YES completion:nil];
+        [presentingViewController presentViewController:messageViewController animated:YES completion:nil];
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectDetailViewController.m]: Can you use self here?

File: CareKit/Connect/OCKConnectViewController.m [hunk 7]
```diff
@@ -212,17 +216,27 @@ - (void)setShowEdgeIndicators:(BOOL)showEdgeIndicators {
     [_tableView reloadData];
 }
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectViewController.m]: Add extra new line.

File: CareKit/Connect/OCKConnectViewController.m [hunk 8]
```diff
@@ -254,4 +268,50 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
+- (void)messageComposeViewController:(MFMessageComposeViewController *)controller didFinishWithResult:(MessageComposeResult)result {
+    [self dismissViewControllerAnimated:YES completion:nil];
+    if (result == MessageComposeResultFailed) {
+        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:OCKLocalizedString(@"ERROR_TITLE", nil)
+                                                                                 message:OCKLocalizedString(@"MESSAGE_SEND_ERROR", nil)
+                                                                          preferredStyle:UIAlertControllerStyleAlert];
+        [self presentViewController:alertController animated:YES completion:nil];
+    }
+}
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectViewController.m]: Add extra new line.

File: CareKit/Connect/OCKConnectViewController.m [hunk 9]
```diff
@@ -254,4 +268,50 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
+- (void)mailComposeController:(MFMailComposeViewController *)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError *)error {
+    [self dismissViewControllerAnimated:YES completion:nil];
+    if (result == MFMailComposeResultFailed) {
+        UIAlertController *alertController = [UIAlertController alertControllerWithTitle:OCKLocalizedString(@"ERROR_TITLE", nil)
+                                                                                 message:OCKLocalizedString(@"EMAIL_SEND_ERROR", nil)
+                                                                          preferredStyle:UIAlertControllerStyleAlert];
+        [self presentViewController:alertController animated:YES completion:nil];
+    }
+}
+
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectViewController.m]: Add extra new line.

File: CareKit/Connect/OCKConnectViewController.m [hunk 10]
```diff
@@ -254,4 +268,50 @@ - (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(N
...
+#pragma mark - UIViewControllerPreviewingDelegate
+
+- (nullable UIViewController *)previewingContext:(id <UIViewControllerPreviewing>)previewingContext viewControllerForLocation:(CGPoint)location {
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectViewController.m]: Remove nullable.

File: CareKit/Connect/OCKConnectDetailViewController.m [hunk 11]
```diff
@@ -153,21 +154,21 @@ - (void)makeCallToNumber:(NSString *)number {
...
     }
 }
 
-- (void)sendEmailToAddress:(NSString *)address {
+- (void)sendEmailToAddress:(NSString *)address withDelegate:(id<MFMailComposeViewControllerDelegate>)emailDelegate usingPresentingViewController:(UIViewController *)presentingViewController {
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectDetailViewController.m]: New naming convention deprecates use of prepositions. Change withDelegate to delegate, and usingPresentingViewController to presentingViewController.

File: CareKit/Connect/OCKConnectDetailViewController.m [hunk 12]
```diff
@@ -153,21 +154,21 @@ - (void)makeCallToNumber:(NSString *)number {
     [[UIApplication sharedApplication] openURL:[NSURL URLWithString:stringURL]];
 }
 
-- (void)sendMessageToNumber:(NSString *)number {
+- (void)sendMessageToNumber:(NSString *)number withDelegate:(id<MFMessageComposeViewControllerDelegate>)messageDelegate usingPresentingViewController:(UIViewController *)presentingViewController {
```
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectDetailViewController.m]: New naming convention deprecates use of prepositions. Change withDelegate to delegate, and usingPresentingViewController to presentingViewController.

**Comment by umerkhan-apple** [comment]
The Care Card 3D Touch works great! 
Minor code changes.

One major bug to address is, when 3D touching on connect cell, the app crashes. Something to do with the titleForSharing method on the delegate.

File: CareKit/CareCard/OCKCareCardViewController.m (hunk 3 above)
Comment (tktsubota) [inline CareKit/CareCard/OCKCareCardViewController.m]: @umerkhan-apple The delegate method calls for `nullable`. Also, what would I return if the if-statement is not satisfied and I want to cancel the peek? Returning `nil` is if the finger is in an invalid place to peek.

**Comment by tktsubota** [comment]
@umerkhan-apple Bug appears to only be reproduced on OCKSample, not OCKTest. Still looking into it.

**Comment by tktsubota** [comment]
@umerkhan-apple Found and fixed the bug and pushed up your suggested changes except for the `nullable` suggestion (see my comment about that).

File: CareKit/CareCard/OCKCareCardViewController.m (hunk 3 above)
Comment (tktsubota) [inline CareKit/CareCard/OCKCareCardViewController.m]: I understand this _is_ possible with Objective-C, but I'm just wondering why I should remove it.

Comment (umerkhan-apple) [inline CareKit/CareCard/OCKCareCardViewController.m]: For our implementation files, we don't include nullability specifiers to encourage code legibility. 
It is just a coding guideline for RK and CK. Not anything particular to the implementation of the code itself.

File: CareKit/Connect/OCKConnectViewController.m (hunk 10 above)
Comment (umerkhan-apple) [inline CareKit/Connect/OCKConnectViewController.m]: For our implementation files, we don't include nullability specifiers to encourage code legibility. 
It is just a coding guideline for RK and CK. Not anything particular to the implementation of the code itself.

**Comment by umerkhan-apple** [comment]
@tktsubota Works like a charm! Great!!

One quick question. Could we add the the phone, message, mail, and share icons to the action sheet that comes up when you 3D Touch on the connect master view?

**Comment by umerkhan-apple** [comment]
@tktsubota Got a little too excited and merged it too soon. Could you make the changes to the nullability specifiers in the implementation files, and then re-do the PR? Sorry--appreciate it!

**Pull request merged**